exp/addons/controllers/clusterresourceset_helpers.go

```diff
@@ -0,0 +1,236 @@
+		if u.Object == nil {
+			continue
+		}
+
+		ret = append(ret, u)
+		count++
+	}
+
+	return ret, nil
+}
+
+func Apply(ctx context.Context, c client.Client, data []byte) error {
+	// Transform the yaml in a list of objects, so following transformation can work on typed objects (instead of working on a string/slice of bytes)
+	reader := bufio.NewReader(bytes.NewReader(data))
+	// TODO: Check how many bytes enough here
+	_, _, isJSON := utilyaml.GuessJSONStream(reader, 2048)
+	if isJSON {
+		var err error
+		data, err = yaml.JSONToYAML(data)
+		if err != nil {
+			return errors.Wrapf(err, "failed converting JSON data to YAML")
+		}
+	}
+	objs, err := ToUnstructured(data)
+	if err != nil {
+		return errors.Wrapf(err, "failed converting data to unstructured objects")
+	}
+
+	sortedObjs := utilresource.SortResourcesForCreate(objs)
+	// TODO: if one of them fails, return with error or continue creating other objects in the resource?
+	for i := range sortedObjs {
+		err = apply(ctx, c, &objs[i])
+		if err != nil {
+			return errors.Wrapf(err,
+				"failed to applying resource %q/%q", objs[i].GetKind(), objs[i].GetName())
+		}
+	}
+	return nil
+}
+
+func apply(ctx context.Context, c client.Client, obj *unstructured.Unstructured) error {
+	// Create the object on the API server.
+	if err := c.Create(ctx, obj); err != nil {
+		// The create call is idempotent, so if the object already exists
+		// then do not consider it to be an error.
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrapf(
+				err,
+				"failed to create object %s %s/%s",
+				obj.GroupVersionKind(),
+				obj.GetNamespace(),
+				obj.GetName())
+		}
+		// If the object already exists, try to update it.
+		// TODO: Should there be an update here? Resource exist but not installed by ClusterResourceSet.
+		if err := c.Update(ctx, obj); err != nil {
+			return errors.Wrapf(
+				err,
+				"failed to update object %s %s/%s",
+				obj.GroupVersionKind(),
+				obj.GetNamespace(),
+				obj.GetName())
+		}
+	}
+	return nil
+}
+
+// GetorCreateClusterResourceSetBinding retrieves ClusterResourceSetBinding resource owned by the cluster or create a new one if not found.
+func GetorCreateClusterResourceSetBinding(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, clusterResourceSet *addonsv1.ClusterResourceSet) (*addonsv1.ClusterResourceSetBinding, error) {
+	clusterResourceSetBinding := &addonsv1.ClusterResourceSetBinding{}
+	clusterResourceSetBindingKey := client.ObjectKey{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name,
+	}
+
+	if err := c.Get(ctx, clusterResourceSetBindingKey, clusterResourceSetBinding); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+
+		clusterResourceSetBinding.Name = cluster.Name
+		clusterResourceSetBinding.Namespace = cluster.Namespace
+		clusterResourceSetBinding.OwnerReferences = util.EnsureOwnerRef(clusterResourceSetBinding.OwnerReferences, metav1.OwnerReference{
```

Comment:
If I got this right, a binding is owned by many clusters, but a ClusterResourceSet has not direct relation with a cluster; if this is true, this means that clusterctl move in its current form won't be capable to detect which ClusterResourceSet should be moved together with a cluster.

Reply:
Working on https://github.com/kubernetes-sigs/cluster-api/pull/3243; my assumption is that that
- ClusterResourceSet owns ClusterResourceSetBinding
- Cluster(s) owns ClusterResourceSetBinding (potentually many)

@sedefsavas is that right?

Reply:
I'm working on https://github.com/kubernetes-sigs/cluster-api/pull/3243 to get it ready for when this PR merges; my assumption is that that
- ClusterResourceSet owns ClusterResourceSetBinding
- Cluster(s) owns ClusterResourceSetBinding (potentually many)

@sedefsavas. @ncdc is that right?

Reply:
Sorry missed this comment. Each cluster has 1 ClusterResourceSetBinding and ClusterResourceSet is also added as owner ref. An example:

```     
apiVersion: addons.cluster.x-k8s.io/v1alpha3
kind: ClusterResourceSetBinding
metadata:
  creationTimestamp: "2020-07-01T06:32:21Z"
  generation: 3
  name: 2test
  namespace: default
  ownerReferences:
  - apiVersion: cluster.x-k8s.io/v1alpha3
    kind: Cluster
    name: 2test
    uid: 90d97990-63fe-4511-9386-9399ea9df5cd
  - apiVersion: addons.cluster.x-k8s.io/v1alpha3
    blockOwnerDeletion: true
    controller: true
    kind: ClusterResourceSet
    name: sedef-crs
    uid: 92af1863-df59-4ff0-896a-003073b198dc
  resourceVersion: "493032"
  selfLink: /apis/addons.cluster.x-k8s.io/v1alpha3/namespaces/default/clusterresourcesetbindings/2test
  uid: 398d93a3-f4c8-41bf-b8f8-c6bcfe99322c
spec:
  bindings:
    sedef-crs:
      resources:
        Secret/sedefjson:
          applied: false
          hash: aDi43qWDYhZoUABitgiOEGf7f2qMRwKVE-C-niAV384=
          lastAppliedTime: "2020-07-01T06:32:22Z"
        Secret/sedefyaml:
          applied: true
          hash: TicmfBWBEPDJRzmsLY5Vk0Ry3GNcN47cMhSDLVkIUh0=
          lastAppliedTime: "2020-07-01T06:32:22Z"

```